bindings/go/src/fdb/fdb.go

```diff
@@ -192,50 +204,98 @@ func MustGetAPIVersion() int {
 }
 
 var apiVersion int
-var networkStarted bool
+var networkStarted, networkStopped bool
 var networkMutex sync.RWMutex
+var networkRunning sync.WaitGroup
 
 var openDatabases sync.Map
 
-func startNetwork() error {
+// executeWithRunningNetworkThread starts the internal network event loop, if not already done,
+// then runs the provided function while network thread is running.
+// This function is safe to be called from multiple goroutines.
+func executeWithRunningNetworkThread(f func()) error {
 	networkMutex.RLock()
+	if networkStopped {
+		networkMutex.RUnlock()
+
+		return ErrNetworkIsStopped
+	}
+
 	if networkStarted {
+
+		// network thread is guaranteed to be running while this user-provided function runs
+		f()
+
 		networkMutex.RUnlock()
 		return nil
 	}
 	// release read lock and acquire write lock
 	networkMutex.RUnlock()
 	networkMutex.Lock()
 	defer networkMutex.Unlock()
-	if networkStarted {
-		return nil
-	}
 
-	if e := C.fdb_setup_network(); e != 0 {
-		return Error{int(e)}
+	if networkStopped {
+		networkMutex.RUnlock()
+
+		return ErrNetworkIsStopped
```

Comment:
```suggestion
	if networkStopped {
		return ErrNetworkIsStopped
```

In this case we have a write mutex and the `defer` will already unlock it. Calling it again will result in a runtime error.

Reply:
Thanks, fixed it! Most likely a copy/paste mistake, my bad.